File: src/org/jetbrains/plugins/scala/compiler/HydraCompilerSettings.java
package org.jetbrains.plugins.scala.compiler;

import com.intellij.openapi.components.PersistentStateComponent;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;
import com.intellij.util.xmlb.XmlSerializerUtil;

import java.util.LinkedList;
import java.util.List;

/**
  * @author Maris Alexandru
  */
@State(
        name = "HydraSettings",
        storages = {@Storage("scala.xml")}
)
class HydraCompilerSettings implements PersistentStateComponent<HydraCompilerSettings> {
  public boolean isHydraEnabled = false;

  public List<String> artifactPaths = new LinkedList<>();

  public String hydraVersion = "";

  public HydraCompilerSettings getState() {
    return this;
  }
  public void loadState(HydraCompilerSettings state) {
    XmlSerializerUtil.copyBean(state, this);
  }

  public static HydraCompilerSettings getInstance() {
    return ServiceManager.getService(HydraCompilerSettings.class);
  }
}

File: src/org/jetbrains/plugins/scala/compiler/HydraCompilerSettings.scala
package org.jetbrains.plugins.scala.compiler

import java.nio.file.Paths

import com.intellij.openapi.project.Project
import org.jetbrains.plugins.scala.compiler.SourcePartitioner.Auto

/**
  * @author Maris Alexandru
  */
class HydraCompilerSettings(project: Project) {

  var noOfCores: String = Math.ceil(Runtime.getRuntime.availableProcessors()/2D).toInt.toString

  var hydraStorePath: String = getDefaultHydraStorePath

  var sourcePartitioner: String = Auto.value

  private def getDefaultHydraStorePath = Paths.get(project.getBaseDir.getPresentableUrl, ".hydra").toString
}

object SourcePartitioner {
  sealed abstract class SourcePartitioner(val value: String)

  case object Auto extends SourcePartitioner("auto")
  case object Explicit extends SourcePartitioner("explicit")
  case object Plain extends SourcePartitioner("plain")
  case object Package extends SourcePartitioner("package")

  val values: Seq[SourcePartitioner] = Seq(Auto, Explicit, Plain, Package)
}

